### PR TITLE
Add resource predicates for tasks

### LINKED ISSUE #694: Add resource predicates for tasks
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**Description**:

Currently, we did not use resource predicates from upstream because of releasing resources; but it makes unnecessary sort and check on the full nodes. It's better to have a predicates to check `Task.Resreq` against `Node.Idle` and `Node.Release` resources.

### PR DESCRIPTION
**Pull request by thandayuthapani**
**What this PR does / why we need it**:
Add resource predicates for tasks
**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/kube-batch/issues/694

### CODE REVIEW

File: pkg/scheduler/plugins/predicates/predicates.go [hunk 1]
```diff
@@ -199,6 +199,13 @@ func (pp *predicatesPlugin) OnSessionOpen(ssn *framework.Session) {
 				node.Name, task.Namespace, task.Name)
 		}
 
+		// Predicate checks whether enough resource is available for the task in that node
+		clonedNode := node.Clone()
+		if !task.Resreq.LessEqual(clonedNode.Idle.Add(node.Releasing)) {
```
Comment (hex108) [inline pkg/scheduler/plugins/predicates/predicates.go]: Do we need use `task.InitResreq` here?

File: pkg/scheduler/plugins/predicates/predicates.go [hunk 2]
```diff
@@ -199,6 +199,13 @@ func (pp *predicatesPlugin) OnSessionOpen(ssn *framework.Session) {
...
+			return fmt.Errorf("task <%s/%s> ResourceFit failed on node <%s>",
+				node.Name, task.Namespace, task.Name)
```
Comment (hex108) [inline pkg/scheduler/plugins/predicates/predicates.go]: `node.Name` should be put at last variable

File: pkg/scheduler/plugins/predicates/predicates.go [hunk 3]
```diff
@@ -199,6 +199,13 @@ func (pp *predicatesPlugin) OnSessionOpen(ssn *framework.Session) {
 				node.Name, task.Namespace, task.Name)
 		}
 
+		// Predicate checks whether enough resource is available for the task in that node
+		clonedNode := node.Clone()
```
Comment (hex108) [inline pkg/scheduler/plugins/predicates/predicates.go]: Just need clone `Idle` resource?

File: pkg/scheduler/plugins/predicates/predicates.go (hunk 1 above)
Comment (k82cn) [inline pkg/scheduler/plugins/predicates/predicates.go]: we also need to check `Releasing` resources.

Comment (thandayuthapani) [reply pkg/scheduler/plugins/predicates/predicates.go]: done

File: pkg/scheduler/plugins/predicates/predicates.go (hunk 2 above)
Comment (thandayuthapani) [reply pkg/scheduler/plugins/predicates/predicates.go]: done

File: pkg/scheduler/plugins/predicates/predicates.go (hunk 3 above)
Comment (thandayuthapani) [reply pkg/scheduler/plugins/predicates/predicates.go]: done

File: pkg/scheduler/plugins/predicates/predicates.go [hunk 4]
```diff
@@ -199,6 +199,13 @@ func (pp *predicatesPlugin) OnSessionOpen(ssn *framework.Session) {
 				node.Name, task.Namespace, task.Name)
 		}
 
+		// Predicate checks whether enough resource is available for the task in that node
+		clonedNode := node.Idle.Clone()
+		if !task.InitResreq.LessEqual(clonedNode.Add(node.Releasing)) {
+			return fmt.Errorf("task <%s/%s> ResourceFit failed on node <%s>",
+				node.Name, task.Namespace, node.Name)
```
Comment (k82cn) [inline pkg/scheduler/plugins/predicates/predicates.go]: should be `task.Namespace, task.Name, node.Name`.

**Comment by thandayuthapani** [comment]
If we add Resource predicate check in predicate plugin, it affects preemption also, so have moved that check to Allocate alone

File: pkg/scheduler/plugins/predicates/predicates.go (hunk 4 above)
Comment (thandayuthapani) [reply pkg/scheduler/plugins/predicates/predicates.go]: done

File: pkg/scheduler/actions/allocate/allocate.go [hunk 5]
```diff
@@ -144,9 +144,16 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
 					glog.V(3).Infof("Predicates failed for task <%s/%s> on node <%s>: %v",
 						task.Namespace, task.Name, node.Name, err)
 				} else {
-					workerLock.Lock()
-					predicateNodes = append(predicateNodes, node)
-					workerLock.Unlock()
+					clonedNode := node.Idle.Clone()
+					// Check for Resource Predicate
+					if task.InitResreq.LessEqual(clonedNode.Add(node.Releasing)) {
```
Comment (k82cn) [inline pkg/scheduler/actions/allocate/allocate.go:149]: let's check this before other predicates; if predicates failed, we can return it directly.

**Comment by k82cn** [comment]
@thandayuthapani , thanks very much for your effort. I resolved conflict at #717 to catch up internal schedule, sorry for that.

**Pull request closed without merging**